### PR TITLE
Use gRPC local channel credentials

### PR DESCRIPTION
**Pull request by JimmyCYJ**
Add a new field local_credentail into GoogleGrpc which supports Envoy to use gRPC local channel credentials.
Updated gRPC library to 1.15.0 release, which provides new methods that we need in order to use local channel credentials. See https://github.com/grpc/grpc/pull/15909.
Certain Google gRPC features, such as passing Google default call credential, only works with a valid channel credential. Local credential is a valid channel credential.

Signed-off-by: JimmyCYJ jimmychen.0102@gmail.com

Risk Level: Low
Testing: Unit test

### CODE REVIEW

**Comment by JimmyCYJ** [comment]
cc @lizan @qiwzhang

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -52,9 +52,9 @@ REPOSITORY_LOCATIONS = dict(
         remote = "https://github.com/google/libprotobuf-mutator",
     ),
     com_github_grpc_grpc = dict(
-        sha256 = "c747e4d903f7dcf803be53abed4e4efc5d3e96f6c274ed1dfca7a03fa6f4e36b",
-        strip_prefix = "grpc-1.14.2",
-        urls = ["https://github.com/grpc/grpc/archive/v1.14.2.tar.gz"],
+        sha256 = "66d1286fe37f9853af42e1e76f8c2265a2e9ec00afd6dd79a5d23e0880bbe874",
+        strip_prefix = "grpc-1.15.0-pre1",
```
Comment (lizan) [inline bazel/repository_locations.bzl]: Do you know when 1.15 will be released?

File: api/envoy/api/v2/core/grpc_service.proto [hunk 2]
```diff
@@ -53,6 +53,11 @@ message GrpcService {
 
         // https://grpc.io/grpc/cpp/namespacegrpc.html#a6beb3ac70ff94bd2ebbd89b8f21d1f61
         google.protobuf.Empty google_default = 2;
+
+        // Use local channel credentials. Currently only UDS is supported.
+        // See
+        // https://github.com/grpc/grpc/blob/782e5c2b8dc6973b6fc0959ba08da0f5b9db2f6f/include/grpcpp/security/credentials.h#L239
+        bool local_credentail = 3;
```
Comment (lizan) [inline api/envoy/api/v2/core/grpc_service.proto]: Shouldn't this be `google.protobuf.Empty`? just like one above. Or define an enum like https://github.com/grpc/grpc/blob/525c59f74cbfe5300ce0c21dbd00fcf7a01e9934/include/grpc/grpc_security_constants.h#L112

File: test/common/grpc/google_grpc_creds_test.cc [hunk 3]
```diff
@@ -20,6 +20,9 @@ TEST(CredsUtility, SslChannelCredentials) {
   EXPECT_EQ(nullptr, CredsUtility::sslChannelCredentials(config));
   creds->mutable_ssl_credentials();
   EXPECT_NE(nullptr, CredsUtility::sslChannelCredentials(config));
+  creds->clear_ssl_credentials();
```
Comment (lizan) [inline test/common/grpc/google_grpc_creds_test.cc]: Not needed as it is oneof.

File: source/common/grpc/google_grpc_creds_impl.cc [hunk 4]
```diff
@@ -11,15 +11,19 @@ namespace Grpc {
...
-  if (google_grpc.has_channel_credentials() &&
-      google_grpc.channel_credentials().has_ssl_credentials()) {
-    const auto& ssl_credentials = google_grpc.channel_credentials().ssl_credentials();
-    const grpc::SslCredentialsOptions ssl_credentials_options = {
-        .pem_root_certs = Config::DataSource::read(ssl_credentials.root_certs(), true),
-        .pem_private_key = Config::DataSource::read(ssl_credentials.private_key(), true),
-        .pem_cert_chain = Config::DataSource::read(ssl_credentials.cert_chain(), true),
-    };
-    return grpc::SslCredentials(ssl_credentials_options);
+  if (google_grpc.has_channel_credentials()) {
```
Comment (lizan) [inline source/common/grpc/google_grpc_creds_impl.cc]: switch(google_grpc.channel_credentials().credentials_specifier_case()) {...

File: bazel/repository_locations.bzl (hunk 1 above)
Comment (JimmyCYJ) [reply bazel/repository_locations.bzl]: It should be 09/11, which is today.

File: source/common/grpc/google_grpc_creds_impl.cc (hunk 4 above)
Comment (JimmyCYJ) [reply source/common/grpc/google_grpc_creds_impl.cc]: Done. Thanks!

File: test/common/grpc/google_grpc_creds_test.cc (hunk 3 above)
Comment (JimmyCYJ) [reply test/common/grpc/google_grpc_creds_test.cc]: Removed. Thanks!

File: api/envoy/api/v2/core/grpc_service.proto (hunk 2 above)
Comment (JimmyCYJ) [reply api/envoy/api/v2/core/grpc_service.proto]: Changed to google.protobuf.Empty.

File: bazel/repository_locations.bzl (hunk 1 above)
Comment (JimmyCYJ) [reply bazel/repository_locations.bzl]: I have changed it to 1.15.0 release now.

File: api/envoy/api/v2/core/grpc_service.proto [hunk 5]
```diff
@@ -53,6 +53,11 @@ message GrpcService {
 
         // https://grpc.io/grpc/cpp/namespacegrpc.html#a6beb3ac70ff94bd2ebbd89b8f21d1f61
         google.protobuf.Empty google_default = 2;
+
+        // Use local channel credentials. Currently only UDS is supported.
+        // See
+        // https://github.com/grpc/grpc/blob/782e5c2b8dc6973b6fc0959ba08da0f5b9db2f6f/include/grpcpp/security/credentials.h#L239
+        google.protobuf.Empty local_credentail = 3;
```
Comment (lizan) [inline api/envoy/api/v2/core/grpc_service.proto]: super nit: local_credentials (to be consistent with gRPC)

Comment (JimmyCYJ) [reply api/envoy/api/v2/core/grpc_service.proto]: Fixed. Thanks.

**Comment by JimmyCYJ** [comment]
cc @htuch

File: source/common/grpc/google_grpc_creds_impl.cc [hunk 6]
```diff
@@ -11,15 +11,23 @@ namespace Grpc {
 
 std::shared_ptr<grpc::ChannelCredentials> CredsUtility::sslChannelCredentials(
```
Comment (lizan) [inline source/common/grpc/google_grpc_creds_impl.cc]: This method need a better name, it is not only about SSL with this change.

Comment (JimmyCYJ) [reply source/common/grpc/google_grpc_creds_impl.cc]: Renamed to getChannelCredentials(). Thanks.

File: api/envoy/api/v2/core/grpc_service.proto [hunk 7]
```diff
@@ -53,6 +53,11 @@ message GrpcService {
 
         // https://grpc.io/grpc/cpp/namespacegrpc.html#a6beb3ac70ff94bd2ebbd89b8f21d1f61
         google.protobuf.Empty google_default = 2;
+
+        // Use local channel credentials. Currently only UDS is supported.
+        // See
+        // https://github.com/grpc/grpc/blob/782e5c2b8dc6973b6fc0959ba08da0f5b9db2f6f/include/grpcpp/security/credentials.h#L239
+        google.protobuf.Empty local_credentials = 3;
```
Comment (htuch) [inline api/envoy/api/v2/core/grpc_service.proto]: This should be a message type, which we make empty for now, to allow us to add new fields when local credentials extends beyond UDS without breaking compat.

File: api/envoy/api/v2/core/grpc_service.proto [hunk 8]
```diff
@@ -53,6 +53,11 @@ message GrpcService {
 
         // https://grpc.io/grpc/cpp/namespacegrpc.html#a6beb3ac70ff94bd2ebbd89b8f21d1f61
         google.protobuf.Empty google_default = 2;
+
+        // Use local channel credentials. Currently only UDS is supported.
+        // See
+        // https://github.com/grpc/grpc/blob/782e5c2b8dc6973b6fc0959ba08da0f5b9db2f6f/include/grpcpp/security/credentials.h#L239
```
Comment (htuch) [inline api/envoy/api/v2/core/grpc_service.proto]: Can you also link to this PR `https://github.com/grpc/grpc/pull/15909` here and also in the PR commit message? Local credentials are not well documented, so this seems the best reference.

Comment (JimmyCYJ) [reply api/envoy/api/v2/core/grpc_service.proto]: This PR is added into comment above googleLocalCredentials, and is added into this PR's description.

File: api/envoy/api/v2/core/grpc_service.proto (hunk 7 above)
Comment (JimmyCYJ) [reply api/envoy/api/v2/core/grpc_service.proto]: Added a new empty message googleLocalCredentials. Thanks.

File: api/envoy/api/v2/core/grpc_service.proto [hunk 9]
```diff
@@ -44,6 +44,11 @@ message GrpcService {
       DataSource cert_chain = 3;
     }
 
+    // Local channel credentials. Only UDS is supported for now.
+    // See https://github.com/grpc/grpc/pull/15909.
+    message googleLocalCredentials {
```
Comment (htuch) [inline api/envoy/api/v2/core/grpc_service.proto]: Nit: `GoogleLocalCredentials`

File: api/envoy/api/v2/core/grpc_service.proto (hunk 7 above)
Comment (quanjielin) [reply api/envoy/api/v2/core/grpc_service.proto]: seems use a dedicated message type hit error https://github.com/envoyproxy/envoy/issues/4578, (I tried to replace ```LocalCredentials``` with ```GoogleDefault```(which is ```types.Empty```), then envoy didn't hit error. 

Could we just use ```types.Empty``` for ```LocalCredentials``` to unblock the integration ?

Comment (htuch) [reply api/envoy/api/v2/core/grpc_service.proto]: @quanjielin I don't quite get why that would happen. `Empty` is itself just an empty message type.

Comment (quanjielin) [reply api/envoy/api/v2/core/grpc_service.proto]: problem solved now(I need to update istio sha to take this PR), thanks !

**Pull request merged**